lib/horde/dynamic_supervisor_impl.ex

```diff
@@ -649,6 +646,49 @@ defmodule Horde.DynamicSupervisorImpl do
     |> Map.put(:local_process_count, new_local_process_count)
   end
 
+  defp handoff_child(child, state) do
+    reply =
+      Horde.ProcessesSupervisor.terminate_child_by_id(
```

Comment:
The supervisor will block on this. I'm not sure this is what we want. We are going to have to think about how to avoid this.

Reply:
So we are going to have to block somewhere, but as it stands now we are going to be processing all of these handoffs sequentially. They should at least be processed in parallel. One way to do this would be to tell `Horde.ProcessSupervisor` about them all at once.

Reply:
Actually the more I think about it, the more my opinion changes on how this should be dealt with.

I think we should have a special exit signal: `{:shutdown, :process_redistribution}`. In this part of the code, we should ONLY send `Process.exit(pid, {:shutdown, :process_redistribution})`. Then, in `Horde.ProcessSupervisor#maybe_restart_child/5`, we should handle this exit reason appropriately, and notify `Horde.SupervisorImpl` (ala `Horde.ProcessSupervisor.remove_child_from_horde/2`), which can then release the process back to the Horde.

Does this make sense? Please let me know if you have any questions. I'm also open to jumping into a voice call with you to help you make sense of this if you're having any problems.

Reply:
And I am also thinking that something similar should be done for graceful shutdown. The way it's implemented now is kind of janky actually, and I think this exposes a nice way to get that done. But that's something for another PR of course.

Reply:
+10 for use of the word "janky" 😆😆😆haven't heard that in a while, made me laugh.

Yeah alright this makes perfect sense; I can see how this ultimately makes for a more solid structure, plus in the meantime the `Horde.DynamicSupervisorImpl` wouldn't be blocked. 

Also, if we're just going to be calling `Process.exit/2` directly then I could clean up all of the parameter-passing that I shoehorned into `Horde.ProcessesSupervisor` so as to pass the exit reason in via `terminate_child_by_id`. 

So yeah dude, I'll get on it this evening; I don't think any of these changes should take very long to make. I feel like this is getting really close to the finish line. Maybe after this one is closed out, down the road I can take a stab at making the graceful shutdown work in a similar way; but you should first finalize the idea and open an issue to track it.

Anyway it'd be nice to chat if I get stuck, but at this point I think I've got this in the bag; I'm @sikanrong on telegram messenger and sikanrong101 on Skype, add me! It would be great to be connected anyway for future collaborations on Horde or possibly also the MerkleMap or DeltaCRDT implementations. 

We just released our own project called [DDRT](https://github.com/windfish-studio/ddrt) based on those Horde dependency packages, give it a look! Imagine Horde but for distributing a really big spatial index for things in constant movement. We are super-thankful for your open source portfolio dude

Reply:
Oh wow you're fast; I just saw that you did already open a ticket to track graceful shutdown refactoring :)

Reply:
Resolved in [692028a](https://github.com/derekkraan/horde/pull/164/commits/692028ab2aed36e5f36148b7ca29be21ff6de286) 

See comments below (in the main thread) about implementation details. Pretty much everything went according to plan.